### PR TITLE
Add events mailer services.

### PR DESCRIPTION
**Pull request by terabytesoftw**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

### CODE REVIEW

File: config/events.php [hunk 1]
```diff
@@ -2,8 +2,13 @@
 
 declare(strict_types=1);
 
+use App\Contact\Event\MessageSent;
+use App\Contact\Event\MessageNotSent;
+use App\Event\HandleMessageNotSent;
```
Comment (samdark) [inline config/events.php:7]: MessageNotSentHandler

File: src/Event/HandleMessageNotSent.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\Event;
```
Comment (BoShurik) [inline src/Event/HandleMessageNotSent.php:5]: Why not in `Contact` namespace?

File: src/Contact/Event/MessageSent.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ */
+final class MessageSent
+{
+    public function addFlash(FlashInterface $flash): void
+    {
+        $flash->add(
+            'is-success',
+            [
+                'header' => 'System mailer notification.',
+                'body' => 'Thanks to contact us, we\'ll get in touch with you as soon as possible.',
```
Comment (samdark) [inline src/Contact/Event/MessageSent.php:20]: That's very specific for the contact form. We can (theoretically, send many different messages).

**Comment by samdark** [comment]
Overall, either the messaging should be universal or there's no need to complicate things.

**Comment by samdark** [comment]
And the thing is that some messages should not result in a flash message.

**Comment by Mister-42** [comment]
It is quite possible I don't see the advantage over events in this case, or a `Mailer` _and_ `ContactMailer` file side by side, but for me this works just the same.

```php
<?php

declare(strict_types=1);

namespace App\Contact;

use Exception;
use Psr\Http\Message\ServerRequestInterface;
use Psr\Log\LoggerInterface;
use Yiisoft\Aliases\Aliases;
use Yiisoft\Mailer\MailerInterface;
use Yiisoft\Session\Flash\FlashInterface;

class Mailer
{
    private Aliases $aliases;
    private FlashInterface $flash;
    private LoggerInterface $logger;
    private MailerInterface $mailer;
    private string $to;

    public function __construct(
        Aliases $aliases,
        FlashInterface $flash,
        LoggerInterface $logger,
        MailerInterface $mailer,
        string $to
    ) {
        $this->aliases = $aliases;
        $this->flash = $flash;
        $this->logger = $logger;
        $this->mailer = $mailer;
        $this->to = $to;
    }

    public function send(Form $form, ServerRequestInterface $request): void
    {
        $message = $this->mailer->compose(
            'contact',
            [
                'content' => $form->getAttributeValue('content'),
            ]
        )
            ->setSubject("Message from website: {$form->getAttributeValue('subject')}")
            ->setFrom([$form->getAttributeValue('email') => $form->getAttributeValue('name')])
            ->setTo($this->to);

        $attachFiles = $request->getUploadedFiles();
        foreach ($attachFiles as $attachFile) {
            foreach ($attachFile as $file) {
                if ($file->getError() === UPLOAD_ERR_OK) {
                    $message->attachContent(
                        (string) $file->getStream(),
                        [
                            'fileName' => $file->getClientFilename(),
                            'contentType' => $file->getClientMediaType(),
                        ]
                    );
                }
            }
        }

        try {
            $message->send();
            $flashMsg = _('Thank you for contacting us. We will respond to you as soon as possible.');
        } catch (Exception $e) {
            $flashMsg = $e->getMessage();
            $this->logger->error($flashMsg);
        } finally {
            $this->flash->add(
                isset($e) ? 'is-danger' : 'is-success',
                [
                    'body' => $flashMsg,
                ],
                true
            );
        }
    }
}
```

**Comment by BoShurik** [comment]
BTW https://github.com/yiisoft/app/issues/87

**Pull request closed without merging**